extensions/ipynb/src/notebookImagePaste.ts

```diff
@@ -0,0 +1,145 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+
+class CopyPasteEditProvider implements vscode.DocumentPasteEditProvider {
+
+	async provideDocumentPasteEdits(
+		_document: vscode.TextDocument,
+		_ranges: readonly vscode.Range[],
+		dataTransfer: vscode.DataTransfer,
+		_token: vscode.CancellationToken
+	): Promise<vscode.DocumentPasteEdit | undefined> {
+
+		// get foundational notebook data
+		const activeNotebook = vscode.window.activeNotebookEditor?.notebook;
+		const activeSelection = vscode.window.activeNotebookEditor?.selection;
+		const currentCell = activeNotebook?.getCells(activeSelection);
+		if (!currentCell || currentCell?.length !== 1) {
+			return undefined;
+		}
+		const initialAttachments = currentCell[0].metadata?.custom?.attachments;
+		const initialize = !initialAttachments ? true : false;
+
+		// get b64 data from paste
+		const dataItem = dataTransfer.get('image/png');
```

Comment:
Also instead of using `get` you may want to iterate through the data transfer to find all relevant images. That's should let you handle the case where the user copies and pastes multiple images